qiskit_machine_learning/neural_networks/estimator_qnn.py

```diff
@@ -0,0 +1,198 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Estimator quantum neural network class"""
+import logging
+from typing import Optional, Sequence, Tuple, Union
+
+import numpy as np
+from qiskit.algorithms.gradients import BaseEstimatorGradient
+from qiskit.circuit import Parameter, QuantumCircuit
+from qiskit.opflow import PauliSumOp
+from qiskit.primitives import BaseEstimator
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+
+import qiskit_machine_learning.optionals as _optionals
+
+from .neural_network import NeuralNetwork
+
+if _optionals.HAS_SPARSE:
+    # pylint: disable=import-error
+    from sparse import SparseArray
+else:
+
+    class SparseArray:  # type: ignore
+        """Empty SparseArray class
+        Replacement if sparse.SparseArray is not present.
+        """
+
+        pass
+
+
+logger = logging.getLogger(__name__)
+
+
+class EstimatorQNN(NeuralNetwork):
+    """A Neural Network implementation based on the Estimator primitive."""
+
+    def __init__(
+        self,
+        estimator: BaseEstimator,
+        circuit: QuantumCircuit,
+        observables: Sequence[Union[BaseOperator, PauliSumOp]],
+        input_params: Optional[Sequence[Parameter]] = None,
+        weight_params: Optional[Sequence[Parameter]] = None,
+        gradient: Optional[BaseEstimatorGradient] = None,
+        input_gradients: bool = False,
+    ):
+        """
+        Args:
+            estimator: The estimator used to compute neural network's results.
+            circuit: The quantum circuit to represent the neural network.
+            observables: The observables for outputs of the neural network.
+            input_params: The parameters that correspond to the input of the network.
+            weight_params: The parameters that correspond to the trainable weights.
+            gradient: The estimator gradient to be used for the backward pass.
+            input_gradients: Determines whether to compute gradients with respect to input data.
+                Note that this parameter is ``False`` by default, and must be explicitly set to
+                ``True`` for a proper gradient computation when using ``TorchConnector``.
+        """
+        self._estimator = estimator
+        self._circuit = circuit
+        self._observables = observables
+        self._input_params = list(input_params) or []
+        self._weight_params = list(weight_params) or []
+        self._gradient = gradient
```

Comment:
I followed the structure of existing `OpflowQNN` and other QNNs, so we can only read `observables`. It seems QNNs are basically immutable. What do you think? @adekusar-drl

Reply:
Personally, I'm in favor of immutability. I'm happy with getters, in general, classes should be designed in a way that it is easier to create them with another configuration rather than updating them. Especially if setters are heavy.